### PR TITLE
Atanas exercise 5

### PR DESCRIPTION
**Pull request by MaykAkifovski**
Our [Code Review](https://github.com/Systems-Development-and-Frameworks/CountryRoads/pull/17#pullrequestreview-560686307)

### CODE REVIEW

File: backend/src/tests/tests.test.js [hunk 1]
```diff
@@ -1,117 +1,65 @@
...
+        const bobPost = new Post('Bob Post');
+        const alicePost = new Post('Alice Post');
+        await Post.save(session, bobPost);
+        await Post.save(session, alicePost);
+
+        const linkBobWithPost = await session.writeTransaction((tx) =>
+            tx.run('MATCH (u: User {id: $id}), (p: Post {title: $title}) CREATE (u)-[:AUTHORED]->(p) RETURN u',
+              { id: bob.id, title: bobPost.title})
+        )
+        const linkAliceWithPost = await session.writeTransaction((tx) =>
```
Comment (medizinmensch) [inline backend/src/tests/tests.test.js:52]: variables are never used

File: backend/src/tests/tests.test.js [hunk 2]
```diff
@@ -120,163 +68,321 @@ describe("query", () => {
...
+        beforeEach(async () => {
+            const bob = new User('Bob', 'bob@mail.com', 'bobby');
+            const { driver } = context();
+            const session = driver.session();
+            await User.save(session, bob);
+
+            const bobPost = new Post('Bob Post');
+            await Post.save(session, bobPost);
+
+            const linkBobWithPost = await session.writeTransaction((tx) =>
```
Comment (medizinmensch) [inline backend/src/tests/tests.test.js:237]: variable never used

File: backend/src/tests/tests.test.js [hunk 3]
```diff
@@ -120,163 +68,321 @@ describe("query", () => {
...
+            await User.save(session, alice);
+
+            const bobPost = new Post('Bob Post');
+            await Post.save(session, bobPost);
+
+            const linkBobWithPost = await session.writeTransaction((tx) =>
+              tx.run('MATCH (u: User {id: $id}), (p: Post {title: $title}) CREATE (u)-[:AUTHORED]->(p) RETURN u',
+                { id: bob.id, title: bobPost.title})
+            )
+            const linkAliceWithPost = await session.writeTransaction((tx) =>
```
Comment (medizinmensch) [inline backend/src/tests/tests.test.js:318]: variables never used

File: backend/src/tests/tests.test.js [hunk 4]
```diff
@@ -120,163 +68,321 @@ describe("query", () => {
...
+                }
+            });
+        })
+
+        it('authenticates users with valid inputs', async () => {
+            const variables = { email: 'alice@mail.com', password: 'alice'};
+            await expect(mutate( { mutation: LOGIN, variables })).resolves.toMatchObject({
+                errors: undefined,
+                data: {
+                   login: expect.any(String)
```
Comment (medizinmensch) [inline backend/src/tests/tests.test.js:160]: Could be more specific for a more narrowed down response to validate the correct response.

File: backend/src/tests/tests.test.js [hunk 5]
```diff
@@ -120,163 +68,321 @@ describe("query", () => {
...
+                }
+            })
+        });
+
+        it('creates a user with valid email and password', async () => {
+            const variables = { name: 'Bob', email: 'bob@mail.com', password: 'bobbyBoy'}
+            await expect(mutate({ mutation: SIGNUP, variables})).resolves.toMatchObject({
+                errors: undefined,
+                data: {
+                    signup: expect.any(String)
```
Comment (medizinmensch) [inline backend/src/tests/tests.test.js:210]: Could be more specific for a more narrowed down response to validate the correct response.

**Review by medizinmensch — APPROVED** [review]
Hi Team!

You are one of the firsts to request a review for Exercise 5. Congrats for the speed!

To sum up:


:star:   For choosing a scenario and writing installation instructions in README.md.
:star:   For explaining WHY you have chosen the scenario in the README.md.
:star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.
:star:   :star: Data created in mutations are persisted and your queries and mutations   still work as expected.
:star: :star: Your software   tests are free of side effects.
:star:   Your database is never left in an invalid state.
:star: You deny the client to   call mutations and queries of your subschema directly.
:star:   For requesting a review and reviewing another team's PR.


You did a great job in refactoring to the neo4j persistent storage! Code is working, readable and seems reliable. 

Two more things:
1) Everybody with access to your Graph-API can get the password hashes of your users. Not good! But it is sadly a bit cumbersome to deactivate this.
2) You could improve your readme a bit by adding the test and serve command so people don't have to look into your package.json.

That is 10/10 :star: never the less!

Nailed it

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -34,7 +34,8 @@ jobs:
     steps:
       # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
       - uses: actions/checkout@v2
-
+      - uses: KarakoA/setup-neo4j@v1.1.1
```
Comment (ilonae) [inline .github/workflows/main.yml:37]: Really nice that you're sharing Actions here!

File: backend/README.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Motivation
+We have chosen the Local Neo4J DB architecture, because:
+* firstly, we think it is more convenient and reliable to use a self-hosted database. In this way, the system is not dependent on third-party services and sources, as it is by the Remote GraphQL API architecture.
+* secondly, in the course Independent Coursework we are currently working on a project for acknowledging plagiarism in text. It will be of a great advantage for us if we have also experience in working with a graph oriented database rather than only with a documents oriented one.
+# Instructions
+First, run `npm install` to install all dependencies specified in `package.json`. This will install the neo4j driver required to connect to the database.
+
+Before running the application, you should spin up a local instance of the database. We've provided a convinience shell script `run-db.sh` for you that does this.
+The script creates a directory called neodata in the current directory tree and binds that directory to a docker volume. This way, you won't lose any data even in the case of a system crash.
```
Comment (ilonae) [inline backend/README.md:9]: :star: For choosing a scenario and writing installation instructions in README.md
:star: For explaining WHY you have chosen the scenario in the README.md

File: backend/src/security/permissions.js [hunk 8]
```diff
@@ -1,43 +1,29 @@
-const { isAuthenticated, isPostUpvoted, isPostWithTitlePresent, passwordIsValid, passwordIsTooShort, emailIsTaken, canSeeEmail} = require("./rules");
-const { shield, and, not, deny, allow,  chain} = require('graphql-shield');
-const { UserInputError } = require('apollo-server');
+const { isAuthenticated, canSeeEmail } = require("./rules");
+const { shield, and, not, deny, allow} = require('graphql-shield');
 
 const permissions = shield({
     Query: {
-      "*": allow,
+      "*": deny,
```
Comment (ilonae) [inline backend/src/security/permissions.js:6]: :star: You deny the client to call mutations and queries of your subschema directly.

File: backend/src/tests/tests.test.js [hunk 9]
```diff
@@ -1,117 +1,65 @@
...
-        expect(posts).toHaveLength(4);
-        expect(posts).toEqual([
-            { title: "Mike's Post 1", author: { name: 'Mike' } },
-            { title: "Mike's Post 2", author: { name: 'Mike' } },
-            { title: "Atanas's Post 1", author: { name: 'Atanas' } },
-            { title: "Atanas's Post 2", author: { name: 'Atanas' } },
-        ]);
-    });
+    requestMock = { headers: {} };
+    await cleanDatabase();
```
Comment (ilonae) [inline backend/src/tests/tests.test.js:22]: I like how you clean up your database before AND after transactions.
:star: :star: - Your software tests are free of side effects.

File: backend/src/tests/tests.test.js [hunk 10]
```diff
@@ -1,117 +1,65 @@
...
+    const { driver } = context();
+    await driver.close();
+})
+
+describe('Query', () => {
+    beforeEach(async () => {
+        const bob = new User('Bob', 'bob@mail.com', 'bobby');
+        const alice = new User('Alice', 'alice@mail.com', 'alice');
+        const { driver } = context();
```
Comment (ilonae) [inline backend/src/tests/tests.test.js:38]: I think you could refactor some parts so they are executed before each query and mutation, overall.

File: backend/src/tests/tests.test.js [hunk 11]
```diff
@@ -120,163 +68,321 @@ describe("query", () => {
...
-
-    it("upvotes a post - not authorized", async () => {
-        const {
-            errors:[ error ]
-        } = await mutate({ mutation: UPVOTE_POST, variables: { title: "Atanas's Post 1" } });
 
-        expect(error.message).toEqual('Not Authorised!');
+        beforeEach(async () => {
+            const alice = new User('Alice', 'alice@mail.com', 'alice');
+            const { driver } = context();
```
Comment (ilonae) [inline backend/src/tests/tests.test.js:122]: Same as above.

File: backend/src/domain/Post.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+  constructor(title) {
+    this.title = title;
+    this.votes = 0;
+  }
+
+  static async save(session, post) {
+    const { records: postRecords } = await session.writeTransaction((tx) =>
+      tx.run("CREATE (p: Post {title: $title, votes: 0}) RETURN p", post)
+    )
```
Comment (ilonae) [inline backend/src/domain/Post.js:11]: :star: :star: Data created in mutations are persisted and your queries and mutations still work as expected.

**Review by ilonae — APPROVED** [review]
Well done guys!

**Pull request merged**